core: extend systemcontracts OnPersist with call to KeyManagement

AnnaShaleva opened the pull request:
Add a stub of KeyManagement contract. Extend system OnPersist call with a call to KeyManagement OnPersistV2 function. Enable it starting from NeoXAMEV fork. Use the same notation (V2) as for Governance OnPersistV2 to keep the naming fork-aligned.

Implement
https://github.com/bane-labs/go-ethereum/pull/312#discussion_r1791494184.

@txhsl, please review. I added a stub for KeyManagement contract, we can merge this PR as is and then you'll replace this stub by your code in #312.

AnnaShaleva committed:
core: extend systemcontracts OnPersist with call to KeyManagement

Add a stub of KeyManagement contract. Extend system OnPersist call
with a call to KeyManagement OnPersistV2 function. Enable it starting
from NeoXAMEV fork. Use the same notation (V2) as for Governance
OnPersistV2 to keep the naming fork-aligned.

Implement
https://github.com/bane-labs/go-ethereum/pull/312#discussion_r1791494184.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

txhsl commented on privnet/seven/genesis_privnet.json:
```diff
@@ -241,6 +241,10 @@
+      "balance": "0",
+      "code": "0x60806040526004361061006e575f3560e01c806384ccebc91161004c57806384ccebc9146100e2578063ad3cb1cc14610109578063f13c03af14610146578063f46754a11461015e575f80fd5b806326bd0c4e146100725780634f1ef286146100ab57806352d1902d146100c0575b5f80fd5b34801561007d575f80fd5b5061008e600862012121608c1b0181565b6040516001600160a01b0390911681526020015b60405180910390f35b6100be6100b936600461059b565b610172565b ...
+    },
+    "0x1212100000000000000000000000000000000009": {
```
I'm afraid it is not an easy copy-paste here. Let me update in #312 and plz rebase this on that.

AnnaShaleva replied:
Well, it's not a copy-past, I actually compiled a KeyManagement stub and checked that it works. But it's even better now when the full KeyManagement contract is merged, I'll update PR soon.

AnnaShaleva replied:
Oh, now I got your idea, I forgot to recompile the Stub contract with updated self address.

AnnaShaleva replied:
Fixed.

txhsl commented on core/systemcontracts/contracts.go:
```diff
@@ -9,8 +9,14 @@ import (
+	governanceABI = `[{"inputs":[{"internalType":"address","name":"target","type":"address"}],"name":"AddressEmptyCode","type":"error"},{"inputs":[],"name":"CandidateExists","type":"error"},{"inputs":[],"name":"CandidateNotExists","type":"error"},{"inputs":[],"name":"CandidateWithdrawNotAllowed","type":"error"},{"inputs":[{"internalType":"address","name":"implementation","type":"address"}],"name":"E ...
+	// keyManagementABI is a partial ABI of system anti-MEV KeyManagement contract,
+	// it contains a minimum required list of methods needed for system interactions.
+	keyManagementABI = `[{"inputs":[],"name":"onPersistV2","outputs":[],"stateMutability":"nonpayable","type":"function"}]`
```
I also expect a simpler ABI for Governance like this, only a `onPersistV2` and some other necessary methods, without `withdrawRegisterFee`, `votedTo` and so on.

If you agree, I can later open a new PR for this, or after dkg integration.

AnnaShaleva replied:
> If you agree, I can later open a new PR for this, or after dkg integration.

I thought about it, but I have some tests that actually use some other governance methods. I'll fix it in the current PR, but a set of methods will be extended a bit compared to the set that is used by dBFT.

AnnaShaleva replied:
Fixed.

AnnaShaleva committed:
systemcontracts: reduce Governance ABI

Keep only necessary methods, the system itself don't need the whole
contract ABI.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>
